<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Predict Ethnicity • planet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Predict Ethnicity">
<meta property="og:description" content="planet">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">planet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cell_composition.html">Predict Placental Cell Composition</a>
    </li>
    <li>
      <a href="../articles/ethnicity.html">Predict Ethnicity</a>
    </li>
    <li>
      <a href="../articles/gestational_age.html">Predict Gestational Age</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/wvictor14/planet/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ethnicity_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Predict Ethnicity</h1>
            
      
      <small class="dont-index">Source: <a href="http://github.com/wvictor14/planet/blob/master/vignettes/ethnicity.Rmd"><code>vignettes/ethnicity.Rmd</code></a></small>
      <div class="hidden name"><code>ethnicity.Rmd</code></div>

    </div>

    
    
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"planet"</span><span class="op">)</span></code></pre></div>
</div>
<div id="example-data" class="section level3">
<h3 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example Data</h3>
<p>For demonstration, I use 24 samples from a placental DNAm dataset from GEO, (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75196">GSE7519</a>), which contains samples collected in an Australian population. The DNA methylation data (in betas) can be accessed with <code><a href="https://rdrr.io/r/utils/data.html">data(plBetas)</a></code> and corresponding sample information from <code><a href="https://rdrr.io/r/utils/data.html">data(plPhenoData)</a></code>. Note that cpgs have been filtered to a random ~10,000 CpGs, plus the CpGs used inall of the models from this package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span> <span class="co"># for data wrangling</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span> <span class="co">#  plotting</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://victor.rbind.io/planet">planet</a></span><span class="op">)</span>

<span class="co"># load example data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">plBetas</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">plPhenoData</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">plBetas</span><span class="op">)</span>
<span class="co">#&gt; [1] 13918    24</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">plPhenoData</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 7</span>
<span class="co">#&gt;   sample_id  sex   disease   gestation_wk ga_RPC ga_CPC ga_RRPC</span>
<span class="co">#&gt;   &lt;fct&gt;      &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 GSM1944936 Male  preeclam~           36   38.5   38.7    38.7</span>
<span class="co">#&gt; 2 GSM1944939 Male  preeclam~           32   33.1   34.2    32.6</span>
<span class="co">#&gt; 3 GSM1944942 Fema~ preeclam~           32   34.3   35.1    33.3</span>
<span class="co">#&gt; 4 GSM1944944 Male  preeclam~           35   35.5   36.7    35.5</span>
<span class="co">#&gt; 5 GSM1944946 Fema~ preeclam~           38   37.6   37.6    36.6</span>
<span class="co">#&gt; 6 GSM1944948 Fema~ preeclam~           36   36.8   38.4    36.7</span></code></pre></div>
</div>
<div id="infer-ethnicity" class="section level3">
<h3 class="hasAnchor">
<a href="#infer-ethnicity" class="anchor"></a>Infer Ethnicity</h3>
<p><strong>Requirements:</strong></p>
<ul>
<li>CpGs and SNPs that are used in <code>pl_infer_ethnicity</code> need to be in your betas <code>data.frame</code>. You can ensure that you have all features with the <code>ethnicityCpGs</code> vector:</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">ethnicityCpGs</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="va">ethnicityCpGs</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">plBetas</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<p><em>Note that if some features are missing, <code>pl_infer_ethnicity</code> will throw a warning, but will still work.</em></p>
<ul>
<li><p>The betas <code>data.frame</code> needs to have <em>samples in columns</em> and <em>CpGs/snps</em> in rows. The rownames must be CpG/rs identifiers.</p></li>
<li><p>If you have IDAT files available, then I recommend normalizing your betas <code>data.frame</code> using the same normalization methods used on the training data: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3627582/"><strong>noob</strong></a> and <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3546795/"><strong>BMIQ</strong></a>. To apply these, run <code><a href="https://rdrr.io/pkg/minfi/man/preprocessNoob.html">minfi::preprocessNoob()</a></code> on an <code>rgset</code> object and then <code>wateRmelon::BMIQ()</code>. This has already applied to the example data.</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predictEthnicity.html">predictEthnicity</a></span><span class="op">(</span><span class="va">plBetas</span><span class="op">)</span>

<span class="co"># show last 8 rows</span>
<span class="va">results</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="fl">8</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 8 x 7</span>
<span class="co">#&gt;   Sample_ID Predicted_ethni~ Predicted_ethni~ Prob_African</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;                   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 GSM19449~ Asian            Asian                0.0123  </span>
<span class="co">#&gt; 2 GSM19449~ Caucasian        Caucasian            0.0157  </span>
<span class="co">#&gt; 3 GSM19449~ Asian            Asian                0.0208  </span>
<span class="co">#&gt; 4 GSM19449~ Caucasian        Caucasian            0.000928</span>
<span class="co">#&gt; 5 GSM19449~ Caucasian        Caucasian            0.00226 </span>
<span class="co">#&gt; 6 GSM19449~ Caucasian        Caucasian            0.00660 </span>
<span class="co">#&gt; 7 GSM19449~ Caucasian        Caucasian            0.00216 </span>
<span class="co">#&gt; 8 GSM19449~ Caucasian        Caucasian            0.00114 </span>
<span class="co">#&gt; # ... with 3 more variables: Prob_Asian &lt;dbl&gt;,</span>
<span class="co">#&gt; #   Prob_Caucasian &lt;dbl&gt;, Highest_Prob &lt;dbl&gt;</span></code></pre></div>
<p><code>predictEthnicity</code> returns probabilities corresponding to each ethnicity for each sample (e.g <code>Prob_Caucasian</code>, <code>Prob_African</code>, <code>Prob_Asian</code>). A final classification is determined in two ways:</p>
<ol style="list-style-type: decimal">
<li><p><code>Predicted_ethnicity_nothresh</code> - returns a classification corresponding to the highest class-specific probability.</p></li>
<li><p><code>Predicted_ethnicity</code> - if the highest class-specific probability is below <code>0.75</code>, then the the sample is assigned an <code>Amibiguous</code> label. This threshold can be adjusted with the <code>threshold</code> argument. Samples with this label might require special attention in downstream analyses.</p></li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">results</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
        x <span class="op">=</span> <span class="va">Prob_Caucasian</span>, y <span class="op">=</span> <span class="va">Prob_African</span>,
        col <span class="op">=</span> <span class="va">Predicted_ethnicity</span>
    <span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org//reference/label_percent.html">percent</a></span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org//reference/label_percent.html">percent</a></span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"P(Caucasian)"</span>, y <span class="op">=</span> <span class="st">"P(African)"</span><span class="op">)</span></code></pre></div>
<p><img src="fig/ethnicity-1.png" title="plot of chunk ethnicity" alt="plot of chunk ethnicity" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">results</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
        x <span class="op">=</span> <span class="va">Prob_Caucasian</span>, y <span class="op">=</span> <span class="va">Prob_Asian</span>,
        col <span class="op">=</span> <span class="va">Predicted_ethnicity</span>
    <span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org//reference/label_percent.html">percent</a></span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org//reference/label_percent.html">percent</a></span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"P(Caucasian)"</span>, y <span class="op">=</span> <span class="st">"P(Asian)"</span><span class="op">)</span></code></pre></div>
<p><img src="fig/ethnicity-2.png" title="plot of chunk ethnicity" alt="plot of chunk ethnicity" style="display: block; margin: auto;"></p>
<p>We can’t compare this to self-reported ethnicity as it is unavailable. But we know these samples were collected in Sydney, Australia, and are therefore likely mostly European with some East Asian ancestries.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">Predicted_ethnicity</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Asian Caucasian </span>
<span class="co">#&gt;         2        22</span></code></pre></div>
<p><strong>A note on adjustment in differential methylation analysis</strong></p>
<p>Because ‘Ambiguous’ samples might have different mixtures of ancestries, it might be inadequate to adjust for them as one group in an analysis of admixed populations (e.g. 50/50 Asian/African should not be considered the same group as 50/50 Caucasian/African). One solution would be to simply remove these samples. Another would be to adjust for the raw probabilities-in this case, use only two of the three probabilities, since the third will be redundant (probabilities sum to 1). If sample numbers are large enough in each group, stratifying downstream analyses by ethnicity might also be a valid option.</p>
</div>
<div id="references" class="section level3">
<h3 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h3>
<ol style="list-style-type: decimal">
<li><a href="https://epigeneticsandchromatin.%20biomedcentral.com/articles/10.1186/s13072-019-0296-3"><strong>Yuan V</strong>, Price EM, Del Gobbo G, Mostafavi S, Cox B, Binder AM, et al. Accurate ethnicity prediction from placental DNA methylation data. Epigenetics &amp; Chromatin. 2019 Aug 9;12(1):51.</a></li>
</ol>
</div>
<div id="session-info" class="section level3">
<h3 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2020-12-18 r79646)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 19042)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=English_Canada.1252 </span>
<span class="co">#&gt; [2] LC_CTYPE=English_Canada.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=English_Canada.1252</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                   </span>
<span class="co">#&gt; [5] LC_TIME=English_Canada.1252    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    parallel  stats     graphics  grDevices utils    </span>
<span class="co">#&gt; [7] datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] planet_0.99.1               tidyr_1.1.2                </span>
<span class="co">#&gt;  [3] dplyr_1.0.2                 EpiDISH_2.7.0              </span>
<span class="co">#&gt;  [5] bumphunter_1.33.0           locfit_1.5-9.4             </span>
<span class="co">#&gt;  [7] iterators_1.0.13            foreach_1.5.1              </span>
<span class="co">#&gt;  [9] Biostrings_2.59.2           XVector_0.31.1             </span>
<span class="co">#&gt; [11] SummarizedExperiment_1.21.1 Biobase_2.51.0             </span>
<span class="co">#&gt; [13] MatrixGenerics_1.3.0        matrixStats_0.57.0         </span>
<span class="co">#&gt; [15] GenomicRanges_1.43.1        GenomeInfoDb_1.27.3        </span>
<span class="co">#&gt; [17] IRanges_2.25.6              S4Vectors_0.29.6           </span>
<span class="co">#&gt; [19] BiocGenerics_0.37.0         ggplot2_3.3.3              </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] nlme_3.1-151              bitops_1.0-6             </span>
<span class="co">#&gt;  [3] bit64_4.0.5               filelock_1.0.2           </span>
<span class="co">#&gt;  [5] progress_1.2.2            httr_1.4.2               </span>
<span class="co">#&gt;  [7] rprojroot_2.0.2           tools_4.1.0              </span>
<span class="co">#&gt;  [9] doRNG_1.8.2               utf8_1.1.4               </span>
<span class="co">#&gt; [11] R6_2.5.0                  mgcv_1.8-33              </span>
<span class="co">#&gt; [13] DBI_1.1.0                 colorspace_2.0-0         </span>
<span class="co">#&gt; [15] withr_2.3.0               tidyselect_1.1.0         </span>
<span class="co">#&gt; [17] prettyunits_1.1.1         bit_4.0.4                </span>
<span class="co">#&gt; [19] curl_4.3                  compiler_4.1.0           </span>
<span class="co">#&gt; [21] cli_2.2.0                 xml2_1.3.2               </span>
<span class="co">#&gt; [23] DelayedArray_0.17.7       labeling_0.4.2           </span>
<span class="co">#&gt; [25] rtracklayer_1.51.3        scales_1.1.1             </span>
<span class="co">#&gt; [27] quadprog_1.5-8            askpass_1.1              </span>
<span class="co">#&gt; [29] rappdirs_0.3.1            stringr_1.4.0            </span>
<span class="co">#&gt; [31] digest_0.6.27             Rsamtools_2.7.0          </span>
<span class="co">#&gt; [33] rmarkdown_2.6             htmltools_0.5.0          </span>
<span class="co">#&gt; [35] pkgconfig_2.0.3           sparseMatrixStats_1.3.2  </span>
<span class="co">#&gt; [37] highr_0.8                 dbplyr_2.0.0             </span>
<span class="co">#&gt; [39] rlang_0.4.9               rstudioapi_0.13          </span>
<span class="co">#&gt; [41] RSQLite_2.2.1             DelayedMatrixStats_1.13.2</span>
<span class="co">#&gt; [43] farver_2.0.3              BiocIO_1.1.2             </span>
<span class="co">#&gt; [45] generics_0.1.0            BiocParallel_1.25.2      </span>
<span class="co">#&gt; [47] RCurl_1.98-1.2            magrittr_2.0.1           </span>
<span class="co">#&gt; [49] GenomeInfoDbData_1.2.4    Matrix_1.3-2             </span>
<span class="co">#&gt; [51] fansi_0.4.1               Rcpp_1.0.5               </span>
<span class="co">#&gt; [53] munsell_0.5.0             lifecycle_0.2.0          </span>
<span class="co">#&gt; [55] stringi_1.5.3             yaml_2.2.1               </span>
<span class="co">#&gt; [57] MASS_7.3-53               zlibbioc_1.37.0          </span>
<span class="co">#&gt; [59] BiocFileCache_1.15.1      grid_4.1.0               </span>
<span class="co">#&gt; [61] blob_1.2.1                crayon_1.3.4             </span>
<span class="co">#&gt; [63] lattice_0.20-41           splines_4.1.0            </span>
<span class="co">#&gt; [65] GenomicFeatures_1.43.3    hms_0.5.3                </span>
<span class="co">#&gt; [67] knitr_1.30                pillar_1.4.7             </span>
<span class="co">#&gt; [69] rjson_0.2.20              rngtools_1.5             </span>
<span class="co">#&gt; [71] codetools_0.2-18          biomaRt_2.47.1           </span>
<span class="co">#&gt; [73] XML_3.99-0.5              glue_1.4.2               </span>
<span class="co">#&gt; [75] evaluate_0.14             vctrs_0.3.6              </span>
<span class="co">#&gt; [77] locfdr_1.1-8              gtable_0.3.0             </span>
<span class="co">#&gt; [79] openssl_1.4.3             purrr_0.3.4              </span>
<span class="co">#&gt; [81] assertthat_0.2.1          xfun_0.19                </span>
<span class="co">#&gt; [83] restfulr_0.0.13           e1071_1.7-4              </span>
<span class="co">#&gt; [85] class_7.3-17              tibble_3.0.4             </span>
<span class="co">#&gt; [87] GenomicAlignments_1.27.2  AnnotationDbi_1.53.0     </span>
<span class="co">#&gt; [89] memoise_1.1.0             ellipsis_0.3.1           </span>
<span class="co">#&gt; [91] here_1.0.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Victor Yuan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
